### PR TITLE
Fixes from pysmt tests

### PR DESCRIPTION
**Pull request by makaimann**
Various bug fixes from running @cdonovick's pysmt backend for smt-switch through the pysmt tests.

### CODE REVIEW

File: btor/src/boolector_solver.cpp [hunk 1]
```diff
@@ -700,7 +722,7 @@ Term BoolectorSolver::apply_prim_op(PrimOp op, Term t0, Term t1, Term t2) const
 
       std::shared_ptr<BoolectorTerm> bt0 =
           std::static_pointer_cast<BoolectorTerm>(t0);
-      result = boolector_apply(btor, &args[0], 1, bt0->node);
+      result = boolector_apply(btor, &args[0], 2, bt0->node);
```
Comment (amaleewilson) [inline btor/src/boolector_solver.cpp]: I'm assuming this is correct, but it's not immediately clear to me why this 1 was changed to a 2.

File: include/term.h [hunk 2]
```diff
@@ -72,6 +73,13 @@ class TermIterBase
 class TermIter
 {
  public:
+  // typedefs for marking as an input iterator
+  typedef Term value_type;
+  typedef std::ptrdiff_t difference_type;
+  typedef Term * pointer;
```
Comment (amaleewilson) [inline include/term.h]: A bonus comment here for what pointer and reference are doing would be nice.

File: btor/src/boolector_solver.cpp (hunk 1 above)
Comment (makaimann) [reply btor/src/boolector_solver.cpp]: Ahh yeah, good question. It is correct -- fixing a copy-and-paste bug from before. This is telling boolector how many arguments are in args and this particular case should be applying a function to 2 arguments.

File: include/term.h (hunk 2 above)
Comment (makaimann) [reply include/term.h]: Good idea, thanks! Gave a link to iterator_traits and what it's used for in 70f04dda050865c2182eaeca6d22ad60013d4f2e. Didn't really explain it in detail, but I think it makes sense to accept it as c++ magic

**Review by amaleewilson — COMMENTED** [review]
Looks good to me

File: include/term.h (hunk 2 above)
Comment (amaleewilson) [reply include/term.h]: It makes more sense now! I hadn't read the inputiterator docs.

**Comment by makaimann** [comment]
Thanks for the review!

**Pull request merged**